src/lib.rs

```diff
@@ -94,14 +99,104 @@ impl GitOid {
         BufReader<R>: std::io::Read,
     {
         let digest = hash_algo.create_digest();
-        match GitOid::generate_git_oid_from_buffer(digest, content, expected_length) {
-            Ok(v) => Ok(GitOid {
+        let v = GitOid::generate_git_oid_from_buffer(digest, content, expected_length)?;
+        Ok(GitOid {
+            hash_algorithm: hash_algo,
+            len: v.0,
+            value: v.1,
+        })
+    }
+
+    /// generate a bunch of `GitOid`s from a bunch of async
+    /// readers for a given algorithm
+    pub async fn new_from_async_readers<R, I>(
+        hash_algo: HashAlgorithm,
+        content: I,
+    ) -> IOResult<Vector<GitOid>>
+    where
+        R: AsyncReadExt + std::marker::Unpin,
+        I: IntoIterator<Item = (R, usize)>,
+    {
```

Comment:
Now that I've read the entire PR I feel slightly stronger that we should consider replacing the tuple here with a dedicated type. This would require forwarding `AsyncRead` impls though, which currently requires using the [pin-project](https://docs.rs/pin-project/latest/pin_project/) crate. But that should be doable by writing something like this:


```rust
#[pin_project]
struct Source<R> {
    #[pin]
    reader: R,
    length: usize,
}

impl<R> Source<R> {
    // name as `new` maybe?
    fn from_reader(reader: R, length: usize) {
        Self { reader, length }
    }

    // maybe also from_file ?
}

impl AsyncRead for Source {
    fn poll_read(self: Pin<&mut Self>, cx: &mut Context<'_>, buf: &mut [u8] ) -> Poll<Result<usize>> {
        let mut this = self.project();
        this.reader.poll_read(cx, buf)
    }
}
```
_I haven't checked this code, but it should compile module perhaps a few `mut` statements here and there._

Reply:
Oooohhhh... I like this construct. I'll make the code changes. :-)